bgpd/bgpd.c

```diff
@@ -3386,14 +3407,40 @@ int bgp_lookup_by_as_name_type(struct bgp **bgp_val, as_t *as, const char *name,
 		bgp = bgp_get_default();
 
 	if (bgp) {
-		*bgp_val = bgp;
+		if (IS_BGP_INSTANCE_HIDDEN(bgp) && *as != AS_UNSPECIFIED)
+			hidden = true;
+		/* Handle AS number change */
 		if (bgp->as != *as) {
+			if (hidden
+			    || CHECK_FLAG(bgp->vrf_flags, BGP_VRF_AUTO)) {
+				if (hidden) {
+					bgp_create(as, name, inst_type, bgp,
+						   hidden);
+					UNSET_FLAG(bgp->flags,
+						BGP_FLAG_INSTANCE_HIDDEN);
+				} else {
+					bgp->as = *as;
+					UNSET_FLAG(bgp->vrf_flags,
+						   BGP_VRF_AUTO);
+				}
+
+				/* Set all peer's local AS with this ASN */
+				for (ALL_LIST_ELEMENTS(bgp->peer, node, nnode,
+						       peer))
+					peer->local_as = *as;
+				*bgp_val = bgp;
+				return BGP_INSTANCE_EXISTS;
+			}
+
 			*as = bgp->as;
-			return BGP_ERR_AS_MISMATCH;
+			return BGP_ERR_INSTANCE_MISMATCH;
 		}
 		if (bgp->inst_type != inst_type)
 			return BGP_ERR_INSTANCE_MISMATCH;
-		return BGP_SUCCESS;
+		if (hidden)
+			bgp_create(as, name, inst_type, bgp, hidden);
+		*bgp_val = bgp;
+		return BGP_INSTANCE_EXISTS;
```

Comment:
Returning this will break a lot of places where we treat non-zero as an error. I fixed exactly this problem in https://github.com/FRRouting/frr/commit/d09328e5991c9f657758921264492825e7081175.